mypy can't decide if ast.Str.s is a string or not, depending on Python 3.7 vs 3.8

* This is a bug (I think)

```python
import ast
from typing import Optional


def eval_join(node: ast.AST) -> Optional[str]:
    if (isinstance(node, ast.Call) and isinstance(node.func, ast.Attribute)
            and isinstance(node.func.value, ast.Str)
            and node.func.attr == 'join'):
        return node.func.value.s    # <-- mypy --strict complains here
    return None
```

When I run mypy --strict on this file on Python 3.8, I get

```
ast_is_a_string.py:9: error: Returning Any from function declared to return "Optional[str]"
Found 1 error in 1 file (checked 1 source file)
```

When I run mypy --strict on this file on Python 3.6 or 3.7, I get no errors.

If I wrap the return value in a `typing.cast(typing.Str, node.func.value.s)`, mypy on 3.8 becomes happy, but mypy on 3.6 or 3.7 complains about redundant cast to "str".

* What is the behavior/output you expect?

I expect mypy to behave the same on Python 3.6 through 3.8.

* What are the versions of mypy and Python you are using?

mypy 0.770, Python 3.6 through 3.8.

* Do you see the same issue after installing mypy from Git master?

Yes

* What are the mypy flags you are using? (For example --strict-optional)

Just --strict.

--- a/stdlib/3/ast.pyi
+++ b/stdlib/3/ast.pyi
@@ -14,9 +14,21 @@ from typing_extensions import Literal
 from _ast import *  # type: ignore
 
 if sys.version_info >= (3, 8):
-    class Num(Constant): ...
-    class Str(Constant): ...
-    class Bytes(Constant): ...
+    class Num(Constant):
+        value: complex
+        kind: Optional[str]
+        # Aliases for value, for backwards compatibility
+        s: Any
+    class Str(Constant):
+        value: str
+        kind: Optional[str]
+        # Aliases for value, for backwards compatibility
+        s: str
+    class Bytes(Constant):
+        value: bytes
+        kind: Optional[str]
+        # Aliases for value, for backwards compatibility
+        s: bytes
     class NameConstant(Constant): ...
     class Ellipsis(Constant): ...
 

--- a/stdlib/3/ast.pyi
+++ b/stdlib/3/ast.pyi
@@ -16,17 +16,12 @@ from _ast import *  # type: ignore
 if sys.version_info >= (3, 8):
     class Num(Constant):
         value: complex
-        kind: Optional[str]
-        # Aliases for value, for backwards compatibility
-        s: Any
     class Str(Constant):
         value: str
-        kind: Optional[str]
         # Aliases for value, for backwards compatibility
         s: str
     class Bytes(Constant):
         value: bytes
-        kind: Optional[str]
         # Aliases for value, for backwards compatibility
         s: bytes
     class NameConstant(Constant): ...

